host/xtest/pkcs11_1000.c

```diff
@@ -3741,3 +3742,863 @@ static void xtest_pkcs11_test_1016(ADBG_Case_t *c)
+							  0xCC))
+				break;
+
+		Do_ADBG_EndSubCase(c, NULL);
+	}
+
+	/* Test out simple successful cases */
+	for (i = 0; i < ARRAY_SIZE(digest_test_patterns); i++) {
+		Do_ADBG_BeginSubCase(c, "Simple digest tests - oneshot - %s",
+				     digest_test_patterns[i].test_name);
+
+		mechanism.mechanism = digest_test_patterns[i].mecha;
+		memset(data, 0xCC, sizeof(data));
+		memset(digest, 0xCC, sizeof(digest));
+		digest_size = sizeof(digest);
+
+		memcpy(data, digest_test_patterns[i].data,
+		       digest_test_patterns[i].data_size);
+		data_size = digest_test_patterns[i].data_size;
+
+		expect_digest = digest_test_patterns[i].digest;
+		expect_digest_size = digest_test_patterns[i].digest_size;
+
+		rv = C_DigestInit(session, &mechanism);
+		if (!ADBG_EXPECT_CK_OK(c, rv))
+			goto out;
+
+		rv = C_Digest(session, data, data_size, digest, &digest_size);
+		if (!ADBG_EXPECT_CK_OK(c, rv))
+			goto out;
+
+		if (!ADBG_EXPECT_BUFFER(c, expect_digest, expect_digest_size,
+					digest,	digest_size))
+			goto out;
+
+		/* Verify that end of buffer is still 0xCC */
+		for (i = expect_digest_size; i < sizeof(digest); i++)
+			if (!ADBG_EXPECT_COMPARE_UNSIGNED(c, digest[i], ==,
+							  0xCC))
+				break;
+
+		Do_ADBG_EndSubCase(c, NULL);
+	}
+
+	/* Test out key updates */
+
+	Do_ADBG_BeginSubCase(c, "Simple digest tests - AES key update - SHA-256");
+
+	/* Login to Test Token */
+	rv = C_Login(session, CKU_USER,	test_token_user_pin,
+		     sizeof(test_token_user_pin));
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	logged_in = true;
+
+	/* Generate AES key */
+	rv = C_GenerateKey(session, &cktest_aes_keygen_mechanism,
+			   digest_generate_aes_object,
+			   ARRAY_SIZE(digest_generate_aes_object),
+			   &key_handle);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	memset(secret_data, 0xCC, sizeof(data));
+	digest_get_secret_value[0].ulValueLen = sizeof(secret_data);
+
+	/* Get value of generated secret for verification purposes */
+	rv = C_GetAttributeValue(session, key_handle, digest_get_secret_value,
+				 ARRAY_SIZE(digest_get_secret_value));
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	secret_data_size = digest_get_secret_value[0].ulValueLen;
+
+	/* Calculate digest with PKCS11 */
+	mechanism.mechanism = CKM_SHA256;
+
+	memset(data, 0xCC, sizeof(data));
+	memset(digest, 0xCC, sizeof(digest));
+	digest_size = sizeof(digest);
+
+	memcpy(data, digest_test_patterns[0].data,
+	       digest_test_patterns[0].data_size);
+	data_size = digest_test_patterns[0].data_size;
+
+	rv = C_DigestInit(session, &mechanism);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	rv = C_DigestUpdate(session, data, data_size);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	rv = C_DigestKey(session, key_handle);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	rv = C_DigestFinal(session, digest, &digest_size);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	/* Verify digest with openssl */
+	mdctx = EVP_MD_CTX_create();
+	if (!mdctx)
```

Comment:
it would be nice to trace the unlikely failures. Maybe change here to:
```c
	mdctx = EVP_MD_CTX_create();
	if (!ADBG_EXPECT_NOT_NULL(mdctx))
		goto out;
```
and at `out:`:
```
out:
	if (!ADBG_EXPECT_NULL(mdctx)) {
		Do_ADBG_Log("Unexpected failure in openssl functions: %d", ret);
		EVP_MD_CTX_destroy(mdctx);
	}
```

Reply:
OK.